shared-libs/cht-datasource/src/input.ts

```
@@ -0,0 +1,270 @@
+import { convertToUnixTimestamp, DataObject, hasField, hasFields, isRecord } from './libs/core';
+import { InvalidArgumentError } from './libs/error';
+
+/**
+ * An input for a contact
+ */
+type ContactInput = DataObject & Readonly<{
+  type: string,
+  name: string,
+  reported_date?: string | number,
+  _id?: string,
+  _rev?: string
+}>;
+
+/**
+ * Builds an input object for creation and update of a contact with
+ * the given fields.
+ * @param data object containing the fields for a contact
+ * @returns the contact input
+ * @throws Error if data is not an object
+ * @throws Error if type is not provided or is empty
+ * @throws Error if name is not provided or is empty
+ * @throws Error if reported_date is not in a valid format.
+ * Valid formats are 'YYYY-MM-DDTHH:mm:ssZ', 'YYYY-MM-DDTHH:mm:ss.SSSZ', or <unix epoch>.
+ */
+export const validateContactInput = (data: unknown): ContactInput => {
+  return validateContactInputNonAssertive(data) as ContactInput;
+};
+
+/** @internal*/
+export const validateContactInputNonAssertive = (data: unknown): Record<string, unknown> => {
+  if (!isRecord(data)) {
+    throw new InvalidArgumentError('Invalid "data": expected an object.');
+  }
+  const input = { ...data };
+  insertReportedDateIfMissing(input);
+  if (!isValidReportedDate(input.reported_date)) {
+    throw new InvalidArgumentError(
+      'Invalid reported_date. Expected format to be ' +
+      '\'YYYY-MM-DDTHH:mm:ssZ\', \'YYYY-MM-DDTHH:mm:ss.SSSZ\', or a Unix epoch.'
+    );
+  }
+  input.reported_date = convertToUnixTimestamp(input.reported_date as string | number);
```

We don't necessarily have to change anything here, but just wanted to call this out in case you had similar feelings.  The tricky part about this validation is that with the remote adapter, it is going to be run twice (once on the client side when the remote adapter code is called, then again running in the api server when the local adapter is called.)  This is true of all the front-line validation we do here in cht-datasource and it is always a tradeoff between failing early vs avoiding duplicate logic.

I think I am fine leaving this how it is, but I would also be okay either moving all this code down to the local adapter OR splitting it so if a valid is provided for `reported_date`, we validate the value here, but if no value is provided, we wait to set one (and/or convert to unix timestamp) until the local adapter...  I feel like this logic is right on the edge of what I would want to possibly duplicate...